### ISSUE TITLE
Marking jobs as done or error.

### ISSUE THREAD

**Issue by adame21** [nats-io/stan.js#136]
Hi dear lib creator,

I am currently building a project that uses nats-streaming as it's queue.

I have some questions about acknowledging job statuses. 

In my subscription, I have a callback that receives a "data" variable, that has all these functions in it, once the "on message" event triggers I receive the message and send it to a different module for processing, and that module eventually returns either true or false, based on that response I'd like to handle the job differently.

Either I acknowledge it as done successfully and at which point I'd like to remove it from the queue, or I acknowledge it as failed, and I'd like to mark it as such so that later on I can perform retries on it or just do whatever.

1. I noticed in my "data" var I have a function called ack(), is that function used to acknowledge a successful job? Edit: I realized it does acknowledge it as done if I use manualAckMode. all good.
2. Is there a built in function to mark a job as failed? Edit: This is basically done with timeout of the ack right?

3. I saw there's "at least once" functionality in this lib, how would I configure it ?
4. Is there anyway to mark a job as "in process" while its being worked on to avoid another processor from taking it thus doing it twice? Edit: This is the most important part for me.

I apologize in advance if these questions are noobish, I had a rough time getting answers from the documentation and the source code alone.

Thanks.

Edited some questions after messing around with it. thanks.

**Comment by aricart** [comment]
Hi There,

NATS Streaming is a message log implementing "At Least Once" semantics. All messages are stored in the server according to the storage policy. Clients simply play back the messages that are in the log.

If your client is durable, (https://github.com/nats-io/stan.js#durable-subscriptions) the server will remember the last message that the client acknowledged, and will resume playback from there. By default messages are auto-acknowledged. See https://github.com/nats-io/stan.js#message-acknowledgements-and-redelivery.


With that background, answers to your questions:

- 1) Yes you have to set set the subscription option `setManualAckMode(true)`, if your processing will take more than a few seconds (I believe default is 30s), you'll need to set the option `setAckWait(millis)`. If for some reason you fail to ack before the wait time, or if the server for some reason doesn't get the ack, you'll see the message again.

- 2) Correct - there's no fail status - it is simply the status of your subscription. The server doesn't know if you got message or completed processing it until you ack it and it receives the ack.

- 3) You don't have to, this is what you just did with the subscription.

- 4) No there's no way of doing that. If you need to have transaction semantics, you may want to consider storing the message in a database or something, where you can compare its id, and decide to ignore it (by 'ack'ing it so you don't see it again).

You really want to take a look at:

https://nats-io.github.io/docs/developer/streaming/

**Comment by adame21** [comment]
Hello,
Thanks for the reply, it really cleared a lot of things and the link to the documentation was very useful.
I still have a small issue that doesent seem to make sense
I have set setManualAckMode to true, and correctly, as long as it hasn't been ack'd and the amount of time set in setAckWait expires the job gets redelivered.
The problem is that when I do ack the message it doesent seem to get removed.
The message gets re-sent every time I restart my application.
I also set opts like this:
````
        const opts = stan.subscriptionOptions();
        opts.setDeliverAllAvailable();
        opts.setManualAckMode(true);
        opts.setAckWait(60 * 1000); //60 seconds.
````
Is the options done wrong? am i supposed to still be getting the message redelivered everytime it starts up because of these options?
What I'm expecting is for the message to not appear again after getting ack'd
just to make sure - I ack the message like this:
````      
  subscription.on('message', (data) => {
                        data.ack();
        });
````
Thanks.
